File: backend/src/main/java/com/proj/sms/service/MaterialService.java
package com.proj.sms.service;

import com.proj.sms.interfaces.MaterialRepository;
import com.proj.sms.models.Material;
import com.proj.sms.models.Material.MaterialType;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
public class MaterialService {

    private final MaterialRepository materialRepository;

    public MaterialService(MaterialRepository materialRepository) {
        this.materialRepository = materialRepository;
    }

    // CREATE
    public Material createMaterial(Material material) {
        // In a real app, you would set the uploader and subject from the security context or request DTO
        return materialRepository.save(material);
    }

    // READ (Single)
    public Material getMaterialById(Long id) {
        return materialRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Material not found with id: " + id));
    }

    // READ (All)
    public List<Material> getAllMaterials() {
        return materialRepository.findAll();
    }

    // READ (By Type)
    public List<Material> getMaterialsByType(MaterialType type) {
        return materialRepository.findByType(type);
    }

    // READ (By Subject ID)
    public List<Material> getMaterialsBySubjectId(Long subjectId) {
        return materialRepository.findBySubjectId(subjectId);
    }

    // READ (By Uploader ID)
    public List<Material> getMaterialsByUploaderId(Long userId) {
        return materialRepository.findByUploadedById(userId);
    }

    // UPDATE
    public Material updateMaterial(Long id, Material materialDetails) {
        Material material = getMaterialById(id);
        material.setTitle(materialDetails.getTitle());
        material.setDescription(materialDetails.getDescription());
        material.setUrl(materialDetails.getUrl());
        material.setType(materialDetails.getType());
        // The subject and uploader are generally not changed after creation
        return materialRepository.save(material);
    }

    // DELETE
    public void deleteMaterial(Long id) {
        if (!materialRepository.existsById(id)) {
            throw new RuntimeException("Material not found with id: " + id);
        }
        materialRepository.deleteById(id);
    }
}

File: frontend/src/services/materialService.js
// src/services/materialService.js
import api from './api';

const MATERIALS_API_URL = '/materials';

export const getAllMaterials = async () => {
    const response = await api.get(MATERIALS_API_URL);
    return response.data;
};

export const getMaterialById = async (id) => {
    const response = await api.get(`${MATERIALS_API_URL}/${id}`);
    return response.data;
};

export const getMaterialsByType = async (type) => {
    const response = await api.get(`${MATERIALS_API_URL}/type/${type}`);
    return response.data;
};

export const getMaterialsBySubjectId = async (subjectId) => {
    const response = await api.get(`${MATERIALS_API_URL}/subject/${subjectId}`);
    return response.data;
};

export const getMaterialsByUploaderId = async (userId) => {
    const response = await api.get(`${MATERIALS_API_URL}/uploader/${userId}`);
    return response.data;
};

export const createMaterial = async (material) => {
    const response = await api.post(MATERIALS_API_URL, material);
    return response.data;
};

export const updateMaterial = async (id, materialDetails) => {
    const response = await api.put(`${MATERIALS_API_URL}/${id}`, materialDetails);
    return response.data;
};

export const deleteMaterial = async (id) => {
    await api.delete(`${MATERIALS_API_URL}/${id}`);
};